ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionHeartbeatService.java

```diff
@@ -0,0 +1,256 @@
+    config.setMaxTotal(5);
+    config.setBlockWhenExhausted(true);
+    config.setMaxWaitMillis(2000);
+    config.setTestOnBorrow(false);
+    config.setTestOnCreate(false);
+    config.setTestOnReturn(false);
+    config.setTestWhileIdle(false);
+    clientPool = new GenericObjectPool<>(new IMetaStoreClientFactory(conf), config);
+    long txnTimeout = MetastoreConf.getTimeVar(conf, MetastoreConf.ConfVars.TXN_TIMEOUT, TimeUnit.MILLISECONDS);
+    initialDelay = txnTimeout / 4;
+    period = txnTimeout / 2;
+  }
+
+  private static final class IMetaStoreClientFactory extends BasePooledObjectFactory<IMetaStoreClient> {
+
+    private final HiveConf conf;
+
+    @Override
+    public IMetaStoreClient create() throws Exception {
+      return HiveMetaStoreUtils.getHiveMetastoreClient(conf);
+    }
+
+    @Override
+    public PooledObject<IMetaStoreClient> wrap(IMetaStoreClient iMetaStoreClient) {
+      return new DefaultPooledObject<>(iMetaStoreClient);
+    }
+
+    @Override
+    public void destroyObject(PooledObject<IMetaStoreClient> p) {
+      p.getObject().close();
+    }
+
+    @Override
+    public boolean validateObject(PooledObject<IMetaStoreClient> p) {
+      //Not in use currently, would be good to validate the client at borrowing/returning, but this needs support from
+      //MetaStoreClient side
+      return super.validateObject(p);
+    }
+
+    public IMetaStoreClientFactory(HiveConf conf) {
+      this.conf = Objects.requireNonNull(conf);
+    }
+
+  }
+
+  private final class CompactionHeartbeater implements Runnable {
+    final private Logger LOG = LoggerFactory.getLogger(CompactionHeartbeater.class);
+    private final long txnId;
+    private final long lockId;
+    private final String tableName;
+    private final Object lock = new Object();
+    private volatile boolean running = false;
+
+
+    @Override
+    public void run() {
+      IMetaStoreClient msc = null;
+      try {
+        synchronized (lock) {
+          running = true;
+        }
+        LOG.debug("Heartbeating compaction transaction id {} for table: {}", txnId, tableName);
+        // Create a metastore client for each thread since it is not thread safe
+        msc = clientPool.borrowObject();
+        msc.heartbeat(txnId, lockId);
+        clientPool.returnObject(msc);
+      } catch (NoSuchElementException nsee) {
+        LOG.error("Compaction transaction heartbeater pool exhausted, unable to heartbeat", nsee);
+      } catch (Exception e) {
+        LOG.error("Error while heartbeating compaction transaction id {} for table: {}", txnId, tableName, e);
+        try {
+          clientPool.invalidateObject(msc);
+        } catch (Exception ex) {
+          LOG.error("Error while invalidating a broken MetaStoreClient instance", e);
+        }
+      } finally {
+        synchronized (lock) {
+          running = false;
+          lock.notifyAll();
+        }
+      }
+    }
+
+    public void waitUntilFinish() throws InterruptedException {
```

Comment:
can't we simply use `join` for this?

Reply:
The executor service hides the executing thread, you have no direct reference for that. You only have the runnable what you pass in, and the Future you get. Therefore it is not possible to join on it.

Reply:
Future has `join` or `get`

Reply:
Join is only in `CompletableFuture<T>`, but not in `ScheduledFuture<T>` which the `ScheduledExecutorService` returns.